jobs/ml-ops/job/ml_training.py

```diff
@@ -0,0 +1,158 @@
+import pandas as pd
+import numpy as np
+from imblearn.over_sampling import SMOTE
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import accuracy_score, precision_score, recall_score, log_loss
+from sklearn.metrics import RocCurveDisplay
+from sklearn.metrics import ConfusionMatrixDisplay
+from sklearn.model_selection import RandomizedSearchCV
+from sklearn.ensemble import RandomForestClassifier
+
+def load_data(path: str) -> pd.DataFrame:
+    data = pd.read_csv(path, sep=';')
+    return data
+
+def clean_data(data: pd.DataFrame) -> pd.DataFrame:
+    """
+    This function removes rows with missing value(s)
+    """
+
+    data = data.dropna()
+    return data
+
+def transform_data(data: pd.DataFrame) -> pd.DataFrame:
+    """
+    This function handles the transformation of categorical variables of the dataset into 0/1 indicators
+    """
+
+    # use the same category for basic education sub-categories
+    data['education'] = np.where(data['education'] =='basic.9y', 'Basic', data['education'])
+    data['education'] = np.where(data['education'] =='basic.6y', 'Basic', data['education'])
+    data['education'] = np.where(data['education'] =='basic.4y', 'Basic', data['education'])
+
+    # transform all categorical variables into 0/1 indicators and remove columns with string categories
+    cat_vars=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']
+    for var in cat_vars:
+        cat_list = 'var' + '_' + var
+        cat_list = pd.get_dummies(data[var], prefix=var)
+        data = data.join(cat_list)
+
+    data_vars = data.columns.values.tolist()
+    to_keep=[i for i in data_vars if i not in cat_vars]
+    data=data[to_keep]
+
+    # normalize column naming
+    data.columns = data.columns.str.replace('.','_')
+    data.columns = data.columns.str.replace(' ','_')
+
+    # replace yes/no by 1/0 for target variable y
+    data['y'] = data['y'].replace(to_replace=['yes', 'no'], value=[1,0])
+ 
+    return data
+
+def split_to_train_test_data(data: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """
+    This function extracts target (predicted) variable and splits data into training/test data
+    """
+
+    # extract target (predicted) variable
+    X = data.loc[:, data.columns != 'y'] # type: ignore
+    y = data.loc[:, data.columns == 'y'] # type: ignore
+    
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=50)
+
+    return X_train, X_test, y_train, y_test
+
+def over_sample_target_class(X_train: pd.DataFrame, y_train: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame]:
+    """
+    This function resamples training data
+    """
```

Comment:
nit: For docstring, you don't need to add `this functions` as a prefix every time. I also recommend inling short docstrings as recommended by this linter: https://github.com/jsh9/pydoclint It's not really important, you can ignore this comment

```suggestion
    """Resamples training data."""
```